soc: arm: stm32: do not include kernel headers in soc.h

ioannisg opened the pull request:
We shall not include core kernel headers in soc.h
header of ARM SoCs. We should try to only include
the vendor headers and auto-generated board header
from DTS.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
soc: arm: do not include kernel headers in soc.h

We shall not include core kernel headers in soc.h
header of ARM SoCs. We should try to only include
the vendor headers and auto-generated board header
from DTS. This commit implements this policy for
the SoCs, whose builds have shown to fail due to
header inclusion cycles.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

nashif commented:
btw, there are more...

```
soc/arm/arm/beetle/soc.h:#include <kernel_includes.h>
soc/arm/atmel_sam/same70/soc.h:#include <kernel_includes.h>
soc/arm/cypress/psoc6/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32f0/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32f1/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32f2/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32f3/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32f4/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32f7/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32g0/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32h7/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32l0/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32l1/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32l4/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32mp1/soc.h:#include <kernel_includes.h>
soc/arm/st_stm32/stm32wb/soc.h:#include <kernel_includes.h>```

ioannisg commented:
> 
> 
> btw, there are more...
> 
> ```
> soc/arm/arm/beetle/soc.h:#include <kernel_includes.h>
> soc/arm/atmel_sam/same70/soc.h:#include <kernel_includes.h>
> soc/arm/cypress/psoc6/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32f0/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32f1/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32f2/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32f3/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32f4/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32f7/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32g0/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32h7/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32l0/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32l1/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32l4/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32mp1/soc.h:#include <kernel_includes.h>
> soc/arm/st_stm32/stm32wb/soc.h:#include <kernel_includes.h>```
> ```

I agree, but they do not create the cycles in all platforms.
I prefer to focus on doing the minimum to fix the failing platforms, and complete the clean-up at later staged.

PS: Something we should have done long ago

ioannisg commented:
@nashif I believe this commit would now fix the CI errors you reported.

pabigot commented:
@ioannisg https://github.com/pabigot/zephyr/tree/pr/19357 has additional patches.  (I don't seem to have permission to push to your branch or to this PR.)

There are four more platforms that need the rework to fix a failure, added as FIXUP patches.

Also a different failure with NXP was revealed: something in the nxp rework changed the order so a HAL header that defined a `LPUART_Type::DATA` field in a peripheral register was included after Zephyr's `<linker/sections.h>` which helpfully provides `#define DATA data` (!), resulting in an "data isn't a field" error.  Showed up on `frdm_kw41z` and maybe others.

In that branch is also a commit that applies the same change to the remaining platforms which by inspection may need it but they haven't been seen to fail (yet).  I'd done that before I saw your comment above about not doing more than the necessary.

Do with them as you wish.

pabigot reviewed (changes requested):
Needs at least a couple additional patches from https://github.com/pabigot/zephyr/tree/pr/19357

ioannisg commented:
@pabigot can you try now? Could it have been that the PR was draft?

galak reviewed (changes requested):
Commit message sys stm32, but has change to NXP driver.

ioannisg commented:
@galak right, thanks. I updated the title.

pabigot commented:
@ioannisg please cherry-pick https://github.com/pabigot/zephyr/commit/ea951b46e91adaafecb78c3cf110185e8737516e for the kw41z failure as it's now failing for a different reason.

Also you need the beetle patch.

ioannisg committed:
drivers: serial: mcux: enforce an error-free header inclusion

We need to change the order of inclusions in
uart_mcux_lpuart.c, to avoid build errors. This
is required since the driver structures contain
a field named DATA, which is also a macro defined
in the linker script.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pabigot commented:
@ioannisg The reason you need to cherry-pick https://github.com/pabigot/zephyr/commit/ea951b46e91adaafecb78c3cf110185e8737516e and remove your NXP change from the stm32 commit is that NXP defines ARRAY_SIZE if it doesn't have it already, while sys/utils.h defines it unconditionally.  So soc.h has to go first, then the NXP header, then devices.h.

kw41z still fails because of this.

pabigot commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -5,6 +5,8 @@
 
 #include <errno.h>
+#include <soc.h>
+#include <fsl_lpuart.h>
```
these were copied, not moved, so they're still below.

ioannisg replied:
Thanks. Trying once more :)

ioannisg replied:
Let's see now

> 
> 
> @ioannisg The reason you need to cherry-pick [pabigot@ea951b4](https://github.com/pabigot/zephyr/commit/ea951b46e91adaafecb78c3cf110185e8737516e) and remove your NXP change from the stm32 commit is that NXP defines ARRAY_SIZE if it doesn't have it already, while sys/utils.h defines it unconditionally. So soc.h has to go first, then the NXP header, then devices.h.
> 
> kw41z still fails because of this.

Right. The reason for my original change was that I wanted the driver structure definition to go before kernel header inclusions, due to the issue with the linker-def macro that had the same name as a variable in a driver struct (see my other comment)

ioannisg commented:
> 
> 
> @ioannisg The reason you need to cherry-pick [pabigot@ea951b4](https://github.com/pabigot/zephyr/commit/ea951b46e91adaafecb78c3cf110185e8737516e) and remove your NXP change from the stm32 commit is that NXP defines ARRAY_SIZE if it doesn't have it already, while sys/utils.h defines it unconditionally. So soc.h has to go first, then the NXP header, then devices.h.
> 
> kw41z still fails because of this.

Yeah, I am aware of that. If you see the original PR that created the problem, https://github.com/zephyrproject-rtos/zephyr/pull/17189/, I had already been playing with the inclusion order, but the order I picked is not proper.

The _original_ issue with this driver, @pabigot is that it defines a struct with a field named DATA, but the linker headers have the following macro definition: `#define DATA data`, and this apparently breaks compilation.

erwango commented:
@ioannisg, I'm fine with the change but not sure (because I didn't checked all versions of this PR)
- why stm32 are stigmatized in the PR title while problem is seen on other platforms :-)
- why you didn't extend fix to all stm32 series (I would be fine if you do)

ioannisg commented:
> @ioannisg, I'm fine with the change but not sure (because I didn't checked all versions of this PR)
> 
> * why stm32 are stigmatized in the PR title while problem is seen on other platforms :-)
> * why you didn't extend fix to all stm32 series (I would be fine if you do)

Thanks! @erwango 
Initially I thought the build errors are present only on stm32 SOCs. Since that's not true, I then re worked the commit message, by the PR title stayed :)

This was a hotfix that need to go asap In, so I focused on the platforms that showed to have build errors. Agreed that we could do a proper clean-up in all Arm SOCs (not only stm32)
